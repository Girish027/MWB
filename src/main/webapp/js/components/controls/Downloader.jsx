import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button } from '@tfs/ui-components';
import Constants from 'constants/Constants';
import * as actionsApp from 'state/actions/actions_app';
import { getLanguage } from 'state/constants/getLanguage';

class Downloader extends Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
    this.downloadAvailableResource = this.downloadAvailableResource.bind(this);
  }

  downloadAvailableResource(file, fileName) {
    // create <a> for downloading the file and then delete it
    const element = document.createElement('a');
    element.style.display = 'none';
    element.href = file;
    element.download = fileName;
    document.body.appendChild(element);
    element.click();
    document.body.removeChild(element);
  }

  onClick(event) {
    event.preventDefault();
    const {
      file,
      fileName,
      fileType,
    } = this.props;
    if (file) {
      this.downloadAvailableResource(file, fileName);
    }
  }

  render() {
    const {
      icon, iconProps = {}, children,
    } = this.props;

    let IconComponent = icon;

    return (
      <span>
        <Button
          type="flat"
          name="download-button"
          key="download-file"
          onClick={this.onClick}
          styleOverride={{
            paddingLeft: '10px',
            fontWeight: 'normal',
            ':focus': {
              outline: 'none',
            },
          }}
        >
          {icon
            ? (<IconComponent {...iconProps} />)
            : null
          }
          {children}
        </Button>
      </span>
    );
  }
}

Downloader.propTypes = {
  children: PropTypes.node,
  file: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Blob),
    PropTypes.instanceOf(File),
  ]),
  fileName: PropTypes.string.isRequired,
  icon: PropTypes.func,
  iconProps: PropTypes.object,
};

Downloader.defaultProps = {
  children: '',
  file: '',
  iconProps: {},
};

export default Downloader;
