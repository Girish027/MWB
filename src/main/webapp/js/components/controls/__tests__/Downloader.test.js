import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Downloader from 'components/controls/Downloader';
import { Download } from '@tfs/ui-components';

describe('<Downloader />', () => {
  const file = new Blob([]);
  const props = {
    file,
    fileName: 'download.file',
    icon: Download,
  };

  let wrapper;

  afterAll(() => {
    jest.clearAllMocks();
  });

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = shallow(<Downloader
        {...props}
      />);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly with just the icon', () => {
      wrapper = shallow(<Downloader
        {...props}
      />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly with just the icon and label', () => {
      wrapper = shallow(<Downloader
        {...props}
      >
        {' '}
Download
        {' '}
      </Downloader>);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly with just label', () => {
      wrapper = shallow(<Downloader
        {...props}
        icon={undefined}
      >
        {' '}
Download
        {' '}
      </Downloader>);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality:', () => {
    let event;
    beforeAll(() => {
      event = {
        preventDefault: jest.fn(),
      };
    });

    beforeEach(() => {
      wrapper = shallow(<Downloader
        {...props}
      />);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    describe('onClick:', () => {
      beforeEach(() => {
        wrapper.instance().downloadAvailableResource = jest.fn();
      });

      test('should prevent default behaviour of triggered event and update state', () => {
        wrapper.instance().onClick(event);
        expect(event.preventDefault).toHaveBeenCalled();
      });

      test('should download file if it is available', () => {
        wrapper.instance().onClick(event);
        expect(wrapper.instance().downloadAvailableResource).toHaveBeenCalledWith(props.file, props.fileName);
      });

      test('should not try to download file if it is not available', () => {
        wrapper.setProps({
          file: undefined,
        });
        wrapper.update();
        wrapper.instance().onClick(event);
        expect(wrapper.instance().downloadAvailableResource).not.toHaveBeenCalled();
      });
    });

    describe('downloadAvailableResource:', () => {
      let click;
      beforeEach(() => {
        click = jest.fn();
        document.createElement = jest.fn().mockImplementationOnce(() => ({
          style: {},
          click,
        }));
        document.body.appendChild = jest.fn();
        document.body.removeChild = jest.fn();
      });

      test('should prevent default behaviour of triggered event and update state', () => {
        wrapper.instance().downloadAvailableResource(props.file, props.fileName);
        expect(document.createElement).toHaveBeenCalledWith('a');
        expect(document.body.appendChild).toHaveBeenCalledWith({
          style: {
            display: 'none',
          },
          href: props.file,
          download: props.fileName,
          click,
        });
        expect(click).toHaveBeenCalled();
        expect(document.body.removeChild).toHaveBeenCalledWith({
          style: {
            display: 'none',
          },
          href: props.file,
          download: props.fileName,
          click,
        });
      });
    });
  });
});
