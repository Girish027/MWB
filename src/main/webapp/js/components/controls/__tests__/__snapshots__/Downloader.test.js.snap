// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Downloader /> Snapshots renders correctly with just label 1`] = `
<span>
  <withTheme(Button)
    key="download-file"
    name="download-button"
    onClick={[Function]}
    styleOverride={
      Object {
        ":focus": Object {
          "outline": "none",
        },
        "fontWeight": "normal",
        "paddingLeft": "10px",
      }
    }
    type="flat"
  >
     
    Download
     
  </withTheme(Button)>
</span>
`;

exports[`<Downloader /> Snapshots renders correctly with just the icon 1`] = `
<span>
  <withTheme(Button)
    key="download-file"
    name="download-button"
    onClick={[Function]}
    styleOverride={
      Object {
        ":focus": Object {
          "outline": "none",
        },
        "fontWeight": "normal",
        "paddingLeft": "10px",
      }
    }
    type="flat"
  >
    <Memo(withTheme(Download)) />
  </withTheme(Button)>
</span>
`;

exports[`<Downloader /> Snapshots renders correctly with just the icon and label 1`] = `
<span>
  <withTheme(Button)
    key="download-file"
    name="download-button"
    onClick={[Function]}
    styleOverride={
      Object {
        ":focus": Object {
          "outline": "none",
        },
        "fontWeight": "normal",
        "paddingLeft": "10px",
      }
    }
    type="flat"
  >
    <Memo(withTheme(Download)) />
     
    Download
     
  </withTheme(Button)>
</span>
`;
